cinder/volume/api.py

```diff
@@ -799,9 +806,40 @@ def initialize_connection(self, context, volume, connector):
             msg = _("The volume connection cannot be initialized in "
                     "maintenance mode.")
             raise exception.InvalidVolume(reason=msg)
-        init_results = self.volume_rpcapi.initialize_connection(context,
-                                                                volume,
-                                                                connector)
+
+        def _migrate_and_initialize_connection():
+            volume_type = {}
+            if volume.volume_type_id:
+                volume_type = volume_types.get_volume_type(
+                    context.elevated(), volume.volume_type_id)
+            request_spec = {
+                'volume_properties': volume,
+                'volume_type': volume_type,
+                'volume_id': volume.id}
+            try:
+                dest = self.scheduler_rpcapi.find_backend_for_connector(
+                    context, connector, request_spec)
+            except exception.NoValidBackend:
+                LOG.exception("The current backend couldn't be used with"
+                              "the provided connector and couldn't find"
+                              "another backend to migrate the volume to.")
+                return None
+            self.volume_rpcapi.migrate_volume(context, volume, dest,
+                                              force_host_copy=False, call=True)
+            volume.refresh()
+            return self.volume_rpcapi.initialize_connection(context, volume,
+                                                            connector)
+        init_results = None
+        try:
+            init_results = self.volume_rpcapi.initialize_connection(context,
+                                                                    volume,
+                                                                    connector)
+        except exception.ConnectorRejected:
+            with excutils.save_and_reraise_exception() as exc_context:
+                if CONF.allow_migration_on_attach:
+                    init_results = _migrate_and_initialize_connection()
```

Comment:
Probably should log an info here to say we got a connector rejected during normal initialize_connection and trying to migrate.